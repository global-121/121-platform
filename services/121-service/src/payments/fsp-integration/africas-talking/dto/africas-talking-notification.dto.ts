import { AfricasTalkingNotificationRequestMetadata } from '@121-service/src/payments/fsp-integration/africas-talking/africas-talking-notification.entity';
import { ApiProperty } from '@nestjs/swagger';
import { IsIn, IsOptional, IsString } from 'class-validator';

export class AfricasTalkingNotificationDto {
  @ApiProperty({ example: 'ATPid_1cbe5df027aca34ff15b885be975bfcd' })
  @IsString()
  public readonly transactionId: string;

  @ApiProperty({ example: 'MobileB2C' })
  @IsString()
  @IsIn(['MobileB2C'])
  public readonly category: string;

  @ApiProperty({ example: 'Athena' })
  @IsString()
  @IsIn(['Mpesa', 'Athena', 'Admin'])
  public readonly provider: string;

  @ApiProperty({ example: 'example' })
  @IsOptional()
  @IsString()
  public readonly providerRefId?: string;

  @ApiProperty({ example: 'example' })
  @IsString()
  public readonly providerChannel: string;

  @ApiProperty({ example: 'example' })
  @IsOptional()
  @IsString()
  public readonly clientAccount?: string;

  @ApiProperty({ example: 'test-product' })
  @IsString()
  public readonly productName: string;

  @ApiProperty({ example: 'Wallet' })
  @IsString()
  @IsIn(['PhoneNumber', 'BankAccount', 'Card', 'Wallet'])
  public readonly sourceType: string;

  @ApiProperty({ example: 'PaymentWallet' })
  @IsString()
  public readonly source: string;

  @ApiProperty({ example: 'PhoneNumber' })
  @IsString()
  @IsIn(['PhoneNumber', 'BankAccount', 'Card', 'Wallet'])
  public readonly destinationType: string;

  @ApiProperty({ example: '+254711123467' })
  @IsString()
  public readonly destination: string;

  @ApiProperty({ example: 'KES 10.00' })
  @IsString()
  public readonly value: string;

  @ApiProperty({ example: 'KES 0.10' })
  @IsOptional()
  @IsString()
  public readonly transactionFee?: string;

  @ApiProperty({ example: 'KES 0.10' })
  @IsOptional()
  @IsString()
  public readonly providerFee?: string;

  @ApiProperty({ example: 'Success' })
  @IsIn(['Success', 'Failed'])
  @IsString()
  public readonly status: string;

  @ApiProperty({ example: 'description' })
  @IsString()
  public readonly description: string;

  @ApiProperty({ example: {} })
  @IsOptional()
  public readonly requestMetadata?: AfricasTalkingNotificationRequestMetadata;

  @ApiProperty({ example: {} })
  @IsOptional()
  public readonly providerMetadata?: Record<string, unknown>;

  @ApiProperty({ example: '2020-09-02' })
  @IsOptional()
  @IsString()
  public readonly transactionDate?: string;
}
