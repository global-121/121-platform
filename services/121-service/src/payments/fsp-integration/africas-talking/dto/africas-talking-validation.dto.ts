import { ApiProperty } from '@nestjs/swagger';
import { IsNumber, IsOptional, IsString } from 'class-validator';

export class AfricasTalkingValidationDto {
  @ApiProperty({ example: 'ATPid_1cbe5df027aca34ff15b885be975bfcd' })
  @IsString()
  public readonly transactionId: string;

  @ApiProperty({ example: '+254711123467' })
  @IsString()
  public readonly phoneNumber: string;

  // This property is in API-documentation, but is not always included in payload in practice
  // @ApiProperty({ example: 'MobileB2C' })
  // @IsString()
  // public readonly category: string;

  @ApiProperty({ example: 'KES' })
  @IsString()
  public readonly currencyCode: string;

  @ApiProperty({ example: 500.0 })
  @IsNumber()
  public readonly amount: number;

  // This property is in API-documentation, but is not always included in payload in practice
  // @ApiProperty({ example: '212.78.193.200' })
  // @IsString()
  // public readonly sourceIpAddress: string;

  @ApiProperty({ example: { programId: '1', payment: '1' } })
  @IsOptional()
  public readonly metadata: Record<string, unknown>;
}
