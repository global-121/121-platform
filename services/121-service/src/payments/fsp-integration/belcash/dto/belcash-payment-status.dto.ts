import { ApiProperty } from '@nestjs/swagger';
import { IsNumber, IsOptional, IsString } from 'class-validator';

export class BelcashPaymentStatusDto {
  @ApiProperty({ example: '123456' })
  @IsString()
  public readonly from: string;

  @ApiProperty({ example: 'Account name' })
  @IsString()
  public readonly fromname: string;

  @ApiProperty({ example: '123456' })
  @IsString()
  public readonly fromaccount: string;

  @ApiProperty({ example: '+251900000000' })
  @IsString()
  public readonly to: string;

  @ApiProperty({ example: 'example' })
  @IsString()
  public readonly toname: string;

  @ApiProperty({ example: 'example' })
  @IsString()
  public readonly toaccount: string;

  @ApiProperty({ example: 25 })
  @IsNumber()
  public readonly amount: number;

  @ApiProperty({ example: 5 })
  @IsNumber()
  public readonly fee: number;

  @ApiProperty({ example: 'ETB' })
  @IsString()
  public readonly currency: string;

  @ApiProperty({ example: '121 program: payment 1' })
  @IsString()
  public readonly description: string;

  @ApiProperty({ example: 'REGULAR_TRANSFER' })
  @IsString()
  public readonly statusdetail: string;

  @ApiProperty({ example: 'LUC000000520482ETH' })
  @IsString()
  @IsOptional()
  public readonly id?: string;

  @ApiProperty({ example: '2021-11-09T11:20:27Z' })
  @IsString()
  @IsOptional()
  public readonly date?: string;

  @ApiProperty()
  @IsString()
  @IsOptional()
  public readonly processdate?: string;

  @ApiProperty()
  @IsString()
  @IsOptional()
  public readonly statuscomment?: string;

  @ApiProperty({ example: 'AUTHORIZING' })
  @IsString()
  public readonly status: string;

  @ApiProperty({
    example: 'ed15b1f7-c579-4a59-8224-56520cd3bdf7-payment-1-1636456825275',
  })
  @IsString()
  @IsOptional()
  public readonly referenceid?: string;

  @ApiProperty({
    example: 'ed15b1f7-c579-4a59-8224-56520cd3bdf7-payment-1-1636456825275',
  })
  @IsString()
  @IsOptional()
  public readonly tracenumber?: string;

  @ApiProperty({ example: 'Lucy' })
  @IsString()
  public readonly system: string;
}
