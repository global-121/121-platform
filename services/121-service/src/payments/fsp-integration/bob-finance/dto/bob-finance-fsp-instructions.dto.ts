export class BobFinanceFspInstructions {
  public 'Receiver First name': string | null;
  public 'Receiver last name': string | null;
  public 'Mobile Number'?: number;
  public 'Email'?: string;
  public 'Amount': number;
  public 'Currency': string;
  public 'Expiry Date'?: string;
}
