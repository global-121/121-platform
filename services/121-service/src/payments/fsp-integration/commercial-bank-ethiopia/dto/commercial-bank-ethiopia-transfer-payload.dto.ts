import { ApiProperty } from '@nestjs/swagger';

export class CommercialBankEthiopiaTransferPayload {
  public debitAmount: number;
  public debitTheIrRef: string;
  public creditTheIrRef: string | null;
  public creditAcctNo: string;
  public creditCurrency: string | null;
  public remitterName: string | null;
  public beneficiaryName: string;
  public status?: string;
}

export class CommercialBankEthiopiaRegistrationData {
  public fieldName: string;
  public value: string;
  public referenceId: string;
}

export class CommercialBankEthiopiaValidationData {
  @ApiProperty()
  public id: number;
  @ApiProperty()
  public fullName: string;
  @ApiProperty()
  public bankAccountNumber: string;
}
