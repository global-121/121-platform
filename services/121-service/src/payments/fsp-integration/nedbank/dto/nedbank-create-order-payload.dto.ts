export interface NedbankCreateOrderPayloadDto {
  Data: {
    Initiation: {
      InstructionIdentification: string;
      InstructedAmount: {
        Amount: string;
        Currency: string;
      };
      DebtorAccount: {
        SchemeName: string;
        Identification: string;
        Name: string;
        SecondaryIdentification: string;
      };
      CreditorAccount: {
        SchemeName: string;
        Identification: string;
        Name: string;
        SecondaryIdentification: string;
      };
    };
    ExpirationDateTime: string;
  };
  Risk: {
    OrderCreateReference: string;
    OrderDateTime: string;
  };
}
