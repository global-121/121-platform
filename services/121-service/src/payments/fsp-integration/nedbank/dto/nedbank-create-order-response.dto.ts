import { NedbankVoucherStatus } from '@121-service/src/payments/fsp-integration/nedbank/enums/nedbank-voucher-status.enum';

// Named in line with https://apim.nedbank.co.za/static/docs/cashout-create-order
// I prefixed the names with Nedbank to avoid any confusion with other services (this is convienent with auto-import and finding files with cmd+p)
// ##TODO: Should we either also prefix interfaces in other FSPs integrations or remove the prefix from this one?
export interface NedbankCreateOrderResponseDto {
  Data: {
    OrderId: string;
    Status: NedbankVoucherStatus;
  };
  Links: {
    Self: string;
  };
  Meta: Record<string, unknown>;
}
