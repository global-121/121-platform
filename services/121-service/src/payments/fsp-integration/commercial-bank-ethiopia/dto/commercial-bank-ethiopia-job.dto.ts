import { PaPaymentDataDto } from '@121-service/src/payments/dto/pa-payment-data.dto';
import { CommercialBankEthiopiaTransferPayload } from '@121-service/src/payments/fsp-integration/commercial-bank-ethiopia/dto/commercial-bank-ethiopia-transfer-payload.dto';

export class CommercialBankEthiopiaJobDto {
  paPaymentData: PaPaymentDataDto;
  paymentNr: number;
  programId: number;
  payload: CommercialBankEthiopiaTransferPayload;
  credentials: { username: string; password: string };
  userId: number;
}
