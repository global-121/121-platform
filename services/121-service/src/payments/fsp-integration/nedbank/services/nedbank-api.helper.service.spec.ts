import { Test, TestingModule } from '@nestjs/testing';

import { CreateOrderResponseNedbankApiDto } from '@121-service/src/payments/fsp-integration/nedbank/dtos/nedbank-api/create-order-response-nedbank-api.dto';
import { ErrorReponseNedbankApiDto } from '@121-service/src/payments/fsp-integration/nedbank/dtos/nedbank-api/error-reponse-nedbank-api.dto';
import { NedbankVoucherStatus } from '@121-service/src/payments/fsp-integration/nedbank/enums/nedbank-voucher-status.enum';
import { NedbankApiHelperService } from '@121-service/src/payments/fsp-integration/nedbank/services/nedbank-api.helper.service';

describe('NedbankApiHelperService', () => {
  let service: NedbankApiHelperService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [NedbankApiHelperService],
    }).compile();

    service = module.get<NedbankApiHelperService>(NedbankApiHelperService);
  });

  describe('isNedbankErrorResponse', () => {
    it('should return true for a valid error response', () => {
      const response: ErrorReponseNedbankApiDto = {
        Message: 'Error',
        Code: 'NB.APIM.Field.Invalid',
        Id: 'some-id',
        Errors: [],
      };

      expect(service.isNedbankErrorResponse(response)).toBe(true);
    });

    it('should return false for a non-error response', () => {
      const response: CreateOrderResponseNedbankApiDto = {
        Data: {
          OrderId: '',
          Status: NedbankVoucherStatus.PENDING,
        },
        Links: { Self: '' },
        Meta: {},
      };
      expect(service.isNedbankErrorResponse(response)).toBe(false);
    });

    it('should return true for a response with no Data property', () => {
      const response = 'Recourse not found';
      expect(service.isNedbankErrorResponse(response)).toBe(true);
    });
  });

  describe('createErrorMessageIfApplicable', () => {
    it('should format an error message correctly', () => {
      const errorResponse: ErrorReponseNedbankApiDto = {
        Message: 'BUSINESS ERROR',
        Code: 'NB.APIM.Field.Invalid',
        Id: '1d3e3076-9e1c-4933-aa7f-69290941ec70',
        Errors: [
          { ErrorCode: 'NB.APIM.Field.Invalid', Message: 'Error message' },
        ],
      };

      const errorMessage =
        service.createErrorMessageIfApplicable(errorResponse);

      expect(errorMessage).toContain('BUSINESS ERROR');
      expect(errorMessage).toContain('NB.APIM.Field.Invalid');
      expect(errorMessage).toContain('Error message');
    });

    it('should return default error for no errors', () => {
      expect(
        service.createErrorMessageIfApplicable({
          Errors: [],
        } as unknown as ErrorReponseNedbankApiDto),
      ).toMatchSnapshot();
    });

    it('should format an error message with only errors', () => {
      const errorResponse: ErrorReponseNedbankApiDto = {
        Message: '',
        Code: '',
        Id: '',
        Errors: [
          { ErrorCode: 'NB.APIM.Field.Invalid', Message: 'Error message' },
        ],
      };

      const errorMessage =
        service.createErrorMessageIfApplicable(errorResponse);

      expect(errorMessage).toContain('Errors: Error message');
      expect(errorMessage).not.toContain('Message:');
      expect(errorMessage).not.toContain('Code:');
      expect(errorMessage).not.toContain('Id:');
    });
  });
});
