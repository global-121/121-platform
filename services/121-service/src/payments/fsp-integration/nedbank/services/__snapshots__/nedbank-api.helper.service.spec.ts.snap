// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NedbankApiHelperService createErrorMessageIfApplicable should return default error for no errors 1`] = `"{"Errors":[]}"`;
